op-gateway/src/OPProofService.ts

```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
+          DISPUTE_GAME_FACTORY_ABI,
+          this.l1Provider
+        );
+      } catch (err) {
+        console.error(err)
+        return null;
+      }
+    }
+
+    if (this.disputeGameFactory) {
+      const respectedGameType = await this.optimismPortal.respectedGameType();
+
+      /**
+       * Get the latest output from the L2Oracle. We're building the proof with this batch
+       * We go a few batches backwards to avoid errors like delays between nodes
+       *
+       */
+      const gameCount = Number(await this.disputeGameFactory.gameCount());
+
+      const games: FindLatestGamesResult[] =
+        await this.disputeGameFactory.findLatestGames(
+          respectedGameType,
+          gameCount - 1,
+          50
+        );
+
+      const timestampNow = (await this.l2Provider.getBlock('latest'))!
+        .timestamp;
+      let disputeGameIndex = -1;
+      let disputeGameLocalIndex = -1;
+
+      for (let i = 0; i < games.length; i++) {
+        if (timestampNow - Number(games[i].timestamp) >= this.delay) {
+          disputeGameIndex = Number(games[i].index);
+          disputeGameLocalIndex = i;
+          break;
+        }
+      }
+
+      // If game is not found then fallback to the L2OutputOracle
+      // for just merged case
+      if (disputeGameIndex != -1) {
+        const [gameType, timestamp, proxy] =
+          await this.disputeGameFactory.gameAtIndex(disputeGameIndex);
+
+        if (
+          gameType != respectedGameType ||
+          timestamp != games[disputeGameLocalIndex].timestamp
+        ) {
+          throw new Error('Mismatched Game Data');
+        }
+
+        const disputeGame = new Contract(
+          proxy,
+          FAULT_DISPUTE_GAME_ABI,
+          this.l1Provider
+        );
+
+        const l2BlockNumber = await disputeGame.l2BlockNumber();
+        const gameStatus = await disputeGame.status();
+
+        // gameStatus == CHALLENGER_WINS
+        if (gameStatus == 1) {
+          throw new DisputeGameChallengedError();
+        }
+
+        return {
+          number: l2BlockNumber,
+          proofType: OPWitnessProofType.DisputeGame,
+          index: disputeGameIndex,
+        };
+      }
+    }
+
+    return null;
+  }
+
+  async findL2OutputIndex(): Promise<OPProvableBlock | null> {
+    if (!this.l2OutputOracle) {
+      try {
+        const l2OutputOracleAddress = await this.optimismPortal.l2Oracle();
+
+        this.l2OutputOracle = new Contract(
+          l2OutputOracleAddress,
+          L2_OUTPUT_ORACLE_ABI,
+          this.l1Provider
+        );
+      } catch (err) {
+        console.error(err)
+        return null;
+      }
+    }
+
+    if (this.l2OutputOracle) {
+      /**
+       * Get the latest output from the L2Oracle. We're building the proof with this batch
+       * We go a few batches backwards to avoid errors like delays between nodes
+       *
+       */
+      const l2OutputIndex =
+        Number(await this.l2OutputOracle.latestOutputIndex()) - this.delay;
```

Comment:
This should also check against the timestamp rather than a number of blocks.

Reply:
Max age is implemented in OPOutputLookup contract I have built at https://github.com/Opti-domains/dispute-game-lookup/blob/main/src/OPOutputLookup.sol. So do L2OutputOracle now using min and max age.